Share vis loader error handling.

lukeelmers opened the pull request:
[Closed in favor of #30594]

Closes #29900 

Fixes issue where errors were not caught and displayed in the UI when using the pipeline data loader.

I pulled the error handling logic out into a function that can be shared between `VisualizeDataLoader` and `PipelineDataLoader`.

Note for reviewers: It is recommended to review the diff with "Hide whitespace changes" selected.

lukeelmers committed:
Share vis loader error handling.

lukeelmers commented on src/ui/public/visualize/loader/visualize_data_loader.ts:
```diff
@@ -117,28 +114,7 @@ export class VisualizeDataLoader {
-      // tslint:disable-next-line
-      console.error(error);
-
-      if (isTermSizeZeroError(error)) {
```
I removed `isTermSizeZeroError` and the corresponding `elasticsearch_errors` directory because it is no longer possible to set a term size: 0 in the UI. And if you add it elsewhere, e.g. URL params, you will get a different error before this line of code even executes.

(This is the only place `isTermSizeZero` error is used)

lukeelmers commented on src/ui/public/visualize/loader/pipeline_data_loader.ts:
```diff
@@ -42,5 +43,11 @@ export class PipelineDataLoader {
       }
     );
+
+    if (pipelineResponse.type === 'error') {
```
`@kbn/interpreter` doesn't actually throw errors when it catches them -- it just gives the pipeline response a `type: 'error'`, hence no `catch` here.

lukeelmers commented on src/ui/public/visualize/loader/errors.ts:
```diff
@@ -17,8 +17,22 @@
+
+  vis.requestError = error;
+  vis.showRequestError =
+    error.type && ['NO_OP_SEARCH_STRATEGY', 'UNSUPPORTED_QUERY'].includes(error.type);
```
I'm not sure if this is necessary either or how to reproduce errors of this type; would welcome any input here.

lukeelmers commented on src/ui/public/visualize/loader/errors.ts:
```diff
@@ -17,8 +17,22 @@
+  console.error(error);
+
+  toastNotifications.addDanger({
+    title: 'Error in visualization',
```
TODO: Need i18n here.

elasticmachine commented:
Pinging @elastic/kibana-app

lukeelmers commented on src/ui/public/visualize/loader/errors.ts:
```diff
@@ -17,8 +17,22 @@
-export function isTermSizeZeroError(error) {
-  return ElasticsearchError.hasRootCause(error, 'size must be positive, got 0');
+import { toastNotifications } from 'ui/notify';
+
```
Unsure if this is the best location for this; thinking it might be better to throw it into a helpers or utils directory, perhaps separating out some of the other pipeline helpers into standalone utils at the same time.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11205/)

lukeelmers commented:
^ appears to be the failing CCR test CJ mentioned earlier. I’ll rebase on latest master tomorrow.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11261/)

ppisljar reviewed (approved):
code LGTM, tested it out on linux chrome and the error messages show up (with pipeline and without)

lukeelmers committed:
Address feedback.

lukeelmers commented on src/ui/public/visualize/loader/visualize_data_loader.ts:
```diff
@@ -71,74 +67,43 @@ export class VisualizeDataLoader {
   public async fetch(params: RequestHandlerParams): Promise<VisResponseData | void> {
-    this.vis.filters = { timeRange: params.timeRange };
-    this.vis.requestError = undefined;
-    this.vis.showRequestError = false;
```
I ended up removing these lines since they could be added in embedded visualize handler, so that all the error-related stuff lives in one place.

lukeelmers commented on src/ui/public/visualize/loader/visualize_data_loader.ts:
```diff
@@ -71,74 +67,43 @@ export class VisualizeDataLoader {
+    if (!canSkipResponseHandler) {
+      this.visData = await Promise.resolve(
+        this.responseHandler(requestHandlerResponse, this.vis.params.dimensions)
+      );
```
Since errors are now caught in embedded visualize handler, the `try/catch` was no longer necessary here.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11393/)

lukeelmers commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11675/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11692/)

sorenlouv committed:
[APM] Show metric-only services in service overview (#30397)

lukeelmers commented:
retest

chandlerprall committed:
Upgrade EUI to 6.10.1 (#30404)

thomasneirynck committed:
[Maps] Fix tileload error messaging / Fix heatmap error message (#30327)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11708/)

lukeelmers commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11753/)

Commit:
[tslint] ban exclusive tests (#30448)

Fixes #30262

Adds tslint rules to explicitly ban `describe.only`, `it.only`, and `test.only`.

nreese committed:
[Maps] upgrade mapboxgl to 0.52.0 (#30431)

Commit:
[yarn] upgrade to @elastic/makelogs 4.4.0 (#30447)

thomasneirynck committed:
[Maps] Correctly handle single-feature joins (#30409)

lukeelmers commented:
yarrrrg. the failed test here is that flaky dashboard embeddable rendering one.

jenkins, test this (please)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11809/)

Commit:
[esArchiver/edit] add command for temporarily extracting archives (#30472)

* [esArchiver/edit] add command for temporarily extracting archives

* [esArchiver/edit] fix method description

* [esArchiver/edit] fix return value doc

* [esArchiver] support editing all or sections of archives

Commit:
skip super flaky test

snide committed:
Make map gradients more subtle (#30345)

Cleanup in the maps app.

ogupte committed:
[APM] Render related errors link in transaction flyout (#29807)

* [APM] closes #29563 by rendering related errors link with error count in transaction flyout

* [APM] improved get_trace query by  narrowing indices and parallelizing queries, improved code org/readability

* [APM] code improvements, split get_trace queries into separate source files

* [APM] remove initial transaction details request in favor of looking up the
current transaction data within the trace (waterfall) data

* Add test for `getWaterfall`

* Revert change to `getWaterfallItems` test

* simplified aggregation, waterfall helpers code, and moved get_trace_errors_per_transaction.ts under 'errors'

* improved naming and readbility of waterfall properties

* removed unused routes and queries and fixed some invisible bugs in the waterfall helpers

* added trace.id in addition to the transaction.id filter in the kuery bar for related errors

jgowdyelastic committed:
[ML] Fixing job deletion slow refresh delay (#30371)

walterra committed:
[ML] Fix double escaping. (#30492)

Because the info tooltip has been migrated to React, the escaping necessary for the previous jQuery based tooltip resulting in a double escaping and special character visible in the tooltip. This fixes it by removing the now unnecessary escaping.

chrisdavies committed:
Make migration mapping change detection more robust (#28252)

Modify the way migrations detect mapping changes. The previous approach simply
diffed the index mappings against the mappings defined in Kibana. The problem was
that sometimes the index mappings will *always* differ. For example, if an index template
is affecting the .kibana* indices. Or if Elasticsearch adds a new magical mapping that
appears in the index, even though not specified (this happened in the 7.0 release). So,
instead of diffing, we now store hashes of our mappings in _meta, and compare against
those.

peteharverson committed:
[ML] Fixes filtering on the metric cards in the data visualizer (#30493)

chrisdavies committed:
Remove custom WebPack bundles from Canvas (#30123)

This removes the custom build process for Canvas plugins. Canvas plugins are now regular
Kibana plugins. They can register server-side functions like so:

```js
server.plugins.interpreter.register({
  serverFunctions: [yourFunctionsHere],
});
```

And client-side functions need to be included in `uiExports.canvas`, then they need to
register themselves like so:

```js
// kbnInterpreter is a global, sadly, but it makes registration simple
kbnInterpreter.register({
  elements: [yourElementsHere],
  browserFunctions: [yourFnsHere],
});
```

jgowdyelastic committed:
[ML] Removing old angular based start datafeed modal (#30392)

* [ML] Removing old and broken start datafeed modal

* fixing typos

* removing need to load job from server

* small refactor

* adding comments

* removing unnecessary async/await

epixa committed:
Bump to 8.0.0

walterra committed:
[ML] Fix Job Audit Messages filter. (#30490)

The way job audit messages were fetched didn't retrieve the expected results if there were deleted jobs with messages still present for these jobs.
This fix allows to specify a list of job IDs to filter the audit messages on. For the jobs list UI, the currently existing job IDs will be passed on to ignore messages from deleted jobs.

walterra committed:
[ML] Removes deprecated $scope.indexPatterns. (#30494)

This was missed during the recent migrations to React of Anomaly Explorer in #28750. It removes the now unnecessary `$scope.indexPatterns`.

nreese committed:
[Maps] do not display mapbox style inspector view unless turned on in kibana.yml (#30428)

* [Maps] do not display mapbox style inspector view unless turned on by kibana.yml

* use better name for config name

tsullivan committed:
[Reporting/PDF/PNG] Hide the global header in browser capture (#30000)

* [Reporting/PDF/PNG] Hide the global header in browser capture

* fix a devastating mistake

tsullivan committed:
Fix/reporting vis preserve (#30456)

* add state field for isPreserveLayoutSupported

* doc updates

ycombinator committed:
Adding breaking changes docs for removed setting (#30505)

rashmivkulkarni committed:
Rollup job Functional UI test  (#30280)

lcawl committed:
[DOCS] Adds placeholders for release notes, breaking changes, highlights (#30545)

tylersmalley committed:
Resets Upgrades Assistant in preperation for 9.0 (#30334)

**Removes APM reindexing**: Since indices created in 6.x will be required to
re-index in 7.x there is no longer a need to support migrating APM indices to
the ECS format since ECS is shipped starting in 7.0

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

legrego committed:
Upgrade grunt to 1.0.3 (#30504)

Commit:
[ftr/services/kbnServer] set uiSetting defaults more often (#30459)

* [ftr/services/kbnServer] set uiSetting defaults more often

* [ftr/services/uiSettings] only init defaults when they're defined

Bargs committed:
Enable query bar datepicker in core apps where appropriate (#30171)

Updates discover, visualize, and dashboard to use the date picker added to the query bar in #29130. There are still some areas where we use the top nav date picker, like Timelion and TSVB which don't have a query bar.

epixa committed:
Skip upgrade assistant test until fixture is updated (#30587)

The test's setup routine loads a data directory into ES that is not
compatible with version 8.0.0, so ES fails to start and the test always
fails.

It is skipped for now to unblock CI, but this should be resolved for 8.0

lukeelmers commented:
Gah, sorry everyone who just got a ping on this. Please ignore -- Botched merge 🙄

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12126/)
